### PR TITLE
Add full ClassProperty support to `transform-flow-comments`

### PR DESCRIPTION
**Pull request by danharper**
**NOTE** Don't merge this. I have some class transform concerns.. Writing them up now.

| Q                 | A
| ----------------- | ---
| Bug fix?          | yes
| Breaking change?  | no
| New feature?      | yes
| Deprecations?     | no
| Spec compliancy?  | no
| Tests added/pass? | yes
| Fixed tickets     | comma-separated list of tickets fixed by the PR, if any
| License           | MIT
| Doc PR            | reference to the documentation PR, if any

Continuing from #4623, this adds support for retaining Class Property type information when `{syntax,transform}-class-properties` is enabled.

Note: I've mentioned in a comment below, but this doesn't add support _with_ `transform-es2015-classes`. (It didn't do that before though, so it's not a removed feature or anything).

e.g.

**Input:**

```js
class X {
  a: number = 2
  b: ?string = null
  c: number
  static d: string
  static e: ?number = 3
}
```

**Current `transform-class-properties` Output:** _(lost type information)_

```js
class X {
  constructor() {
    this.a = 2;
    this.b = null;
  }

}
X.e = 3;
```

**New `transform-class-properties` Output:**

```js
class X {
  constructor() {
    this.a = 2;
    this.b = null;
  } /*:: a: number*/

  /*:: b: ?string*/ /*:: c: number*/
  /*:: static e: ?number*/ /*:: static d: string*/
}
X.e = 3;
```

_I can't work out how to retain the annotations at the top, or why they join on one line, but it works._

**Current `syntax-class-properties` Output:** (values get lost from JS)

```js
class X {
  a /*: number*/ = 2;
  b /*: ?string*/ = null;
  /*:: c: number*/
  /*:: static d: string*/
  static e /*: ?number*/ = 3;
}
```

**New `syntax-class-properties` Output:**

```js
class X {
  a = 2; /*:: a: number*/
  b = null; /*:: b: ?string*/
  /*:: c: number*/
  /*:: static d: string*/
  static e = 3; /*:: static e: ?number*/
}
```

^^ It's a little annoying how the annotation comments are no longer inline. I can fix that, but it breaks `transform-class-properties` by the keeping the inline annotation inside the _constructor_ (which is invalid Flow code):

```js
// invalid
class X {
  /*:: c: number*/
  constructor() {
    this.a /*: number*/ = 2;
    this.b /*: ?string*/ = null;
  }
}
```

### CODE REVIEW

File: packages/babel-plugin-transform-flow-comments/src/index.js [hunk 1]
```diff
@@ -4,8 +4,21 @@ export default function ({ types: t }) {
...
   }
 
+  // https://github.com/babel-plugins/babel-plugin-flow-comments/pull/6#issuecomment-122709012
+  function escapeSource(source) {
+    return source.replace(/\*-\//g, "*-ESCAPED/").replace(/\*\//g, "*-/");
+  }
+
+  function getSource(path) {
+    if (path.isClassProperty()) {
+      return (path.node.static ? "static " : "") + path.get("key").getSource() + path.get("typeAnnotation").getSource();
```
Comment (danharper) [inline packages/babel-plugin-transform-flow-comments/src/index.js]: Having to special-case `ClassProperty` as `path.getSource()` would print out the _value_ of the node too, which we shouldn't duplicate really.

e.g.

```js
class X {
  a = 3; /*:: a: number = 3*/
}
```

**Comment by danharper** [comment]
I'm gonna add some tests for this covering class transforms too, just to be sure.

**Comment by danharper** [comment]
I'm kind of unsure what to do now. `ClassProperty` comments created in this transform get lost in `transform-es2015-classes`. However if that were to be fixed in the future, you'd get invalid Flow code as a result e.g.

```js
// input
class X {
  foo: string = 'bar'
  a(): number {
    return 2
  }
}

// possible, invalid output
var X = function () {
  function X() {
    this.foo/*: string*/ = 'bar';  // <--- Flow syntax error
  }

  _createClass(X, [{
    key: "a",
    value: function a()/*: number*/ { return 2; }
  }])

  return X;
}();
```

It seems constructor functions have to be annotated [in a completely different way](https://github.com/facebook/flow/commit/618186ad5ceb66cf7bf009b2591b74ed885d69bd), something like this:

```js
var X = function () {
  function X() {
    this.foo = 'bar';
  }

  _createClass(X, [{
    key: "a",
    value: function a()/*: number*/ { return 2; }
  }])

  /*:: declare class _X {
    static(): void;
    foo: string;
  }
  X = (X: Class<_X>); */

  return X;
}();
```

_It would also need to add all class methods to `declare class`, and somehow handle `extends`. And the only way it could achieve this is to know if `transform-es2015-classes` is going to run.._

---

Perhaps this transform could have an option like `include-class-properties`? This would then be an explicit opt-in that you're not using `transform-es2015-classes`.

I'm just going to close this, it feels like a rabbit role. I was originally thinking this was the "preferred" way of shipping Flow annotations with your compiled JS, but it's not (copying source file alongside output as `.js.flow` is).

**Comment by danez** [comment]
Thanks for looking into this.

**Pull request closed without merging**